state-chain/pallets/cf-elections/src/electoral_systems/blockchain/delta_based_ingress.rs

```diff
@@ -96,7 +97,65 @@ where
 
 		Ok(())
 	}
+
+	/// Refreshes the election for the given identifier with updated properties if `state`
+	/// is different from the election properties.
+	///
+	/// NOTE: This assumes that the accounts in the properties are a superset of the accounts in the
+	/// state. This holds because whenever we  
+	///
+	/// NOTE: This does not check whether the election has consensus or not, because if the state
+	/// does not reflect the properties this means that we did have consensus at some point,
+	/// and anyways would like to recreate from that state.
+	pub fn recreate_election_if_state_changed<
+		ElectionAccess: ElectionWriteAccess<ElectoralSystem = Self>,
+	>(
+		election_identifier: ElectionIdentifier<<Self as ElectoralSystem>::ElectionIdentifierExtra>,
+		election_access: &mut ElectionAccess,
+		state: <Self as ElectoralSystem>::ElectionState,
+	) -> Result<(), CorruptStorageError> {
+		let properties = election_access.properties()?;
+
+		ensure!(
+			state.keys().all(|account| properties.contains_key(account)),
+			CorruptStorageError::new()
+		);
+
+		// if the keys in state and channels don't match up, or if the balances don't match up
+		// then we refresh the election
+		if state.iter().any(|(account, state_balance)| match properties.get(account) {
+			Some((_, property_balance)) => state_balance.amount != property_balance.amount,
+			None => true,
```

Comment:
Can we ever reach this condition? Isn't it excluded by the `ensure!` check above? (not that it matters much...)

Reply:
Yes, it was supposed to be catched by the case above, and it would have been possible to simply `None => return Err(..)`.

Anyways, I have rewritten quite a bit now.